content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,172 @@
+```{r, error = TRUE}
+say_hello <- function(name) {
+  stopifnot("`name` must be a character." = is.character(name))
+  paste("Hello", name)
+}
+
+say_hello(404)
+```
+
+This is clearly a really great improvement to the functionality of base R. However, we can see from this example that we could create the error message programmatically based on the contents of the test.
+Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
+This way, you don't have to repeat yourself which is generally a good aim [^2].
+This becomes necessary when you start having many input checks in your function or in your package.
+
+[^2]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:
+
+```{r}
+say_hello <- function(name) {
+  # Among other things, check_string() checks that we provide a 
+  # character object of length one
+  checkmate::assert_string(name)
+  paste("Hello", name)
+}
+```
+
+```{r, error = TRUE}
+say_hello(404)
+```
+
+```{r, error = TRUE}
+say_hello(c("Bob", "Alice"))
+```
+
+### Other packages to check function inputs
+
+Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
+We will not get into the details of all of the available options here but below is a list of some of the them. If interested in understanding the various approaches to input taking the documentation
+for these package is a great place to start.
+
+-   testthat
+-   assertthat
+-   check
+-   assertr
+-   assertive
+-   ensurer
+-   `vctrs::vec_assert()`
+
+## What about the future?
+
+In this post, we have discussed some methods to check function inputs, and to generate more informative error messages when doing so.
+However, this always comes with a performance cost, even though it's often relatively limited.
+Zero-cost assertions, as found in some other languages, would require some kind of typing system which R does not currently support.
+Interestingly several other languages have evolved to have typing systems as they have developed (TypeScript as an extension of JavaScript, type annotations in Python).
```

Comment:
Thanks all for this great post! One very minor comment: That final sentence is crucial to understanding broader implications of what you've set out, and would work better without the refs to js and python in parentheses, and perhaps slightly tweaking to make more active. Maybe something like:
> ... typing systems as they have developed. Typescript developed as an extension of JavaScript, and type annotations are now possible in python. 

I like the final link! :smirk:

Reply:
https://github.com/r-hub/blog/pull/150/commits/e5e9394c039f1b0b79c129b2799236a0250ea4cb